[FEATURE] Document TYPO3_CONF_VARS (#1408)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/GlobalValues/Typo3ConfVars/SYS.rst

.. include:: /Includes.rst.txt

.. index::
   TYPO3_CONF_VARS; SYS
   TYPO3_CONF_VARS SYS
.. _typo3ConfVars_sys:

==================================
$GLOBALS['TYPO3_CONF_VARS']['SYS']
==================================

.. index::
   TYPO3_CONF_VARS SYS; fileCreateMask
.. _typo3ConfVars_sys_fileCreateMask:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['fileCreateMask']
====================================================

.. confval:: fileCreateMask

   :type: text
   :Default: 0664

   File mode mask for Unix file systems (when files are uploaded/created).

.. index::
   TYPO3_CONF_VARS SYS; folderCreateMask
.. _typo3ConfVars_sys_folderCreateMask:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['folderCreateMask']
======================================================

.. confval:: folderCreateMask

   :type: text
   :Default: 2775

>    As above, but for folders.

Reviewer: ```suggestion
   Folder mode mask for Unix file systems (when folders are uploaded/created).
```

   This makes sense in all cases where the webserver is running with a
   different user/group as you do. Create a new group on your system and add
   you and the webserver user to the group. Now you can safely set the last
   bit in fileCreateMask/folderCreateMask to 0 (for example 770). Important: The
   user who is running your webserver needs to be a member of the group you
   specify here! Otherwise you might get some error messages.

.. index::
   TYPO3_CONF_VARS SYS; sitename
.. _typo3ConfVars_sys_sitename:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['sitename']
==============================================

.. confval:: sitename

   :type: text
   :Default: 'TYPO3'

   Name of the base-site.

.. index::
   TYPO3_CONF_VARS SYS; encryptionKey
.. _typo3ConfVars_sys_encryptionKey:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['encryptionKey']
===================================================

.. confval:: encryptionKey

   :type: text
   :Default: ''

   This is a "salt" used for various kinds of encryption, CRC checksums and
   validations. You can enter any rubbish string here but try to keep it
   secret. You should notice that a change to this value might invalidate

>    temporary information, URLs etc. At least, clear all cache if you change

Reviewer: ```suggestion
   temporary information, URLs etc. At least, clear all caches if you change
```

   this so any such information can be rebuilt with the new key.

.. index::
   TYPO3_CONF_VARS SYS; cookieDomain
.. _typo3ConfVars_sys_cookieDomain:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['cookieDomain']
==================================================

.. confval:: cookieDomain

   :type: text
   :Default: ''

>    Restricts the domain name for FE and BE session cookies. When setting the

Reviewer: ```suggestion
   Restricts the domain name for frontend and backend session cookies. When setting the
```

>    value to ".domain.com" (replace domain.com with your domain!), login

Reviewer: ```suggestion
   value to ".domain.com" (replace domain.com with your domain), login
```

   sessions will be shared across subdomains. Alternatively, if you have more
   than one domain with sub-domains, you can set the value to a regular
   expression to match against the domain of the HTTP request.

>    The result of the match is used as the domain for the cookie. for example :
>    php:`/\.(example1|example2)\.com$/` or :php:`/\.(example1\.com)|(example2\.net)$/`.
>    Separate domains for FE and BE can be set using

Reviewer: ```suggestion
   The result of the match is used as the domain for the cookie. for example:
   php:`/\.(example1|example2)\.com$/` or :php:`/\.(example1\.com)|(example2\.net)$/`.
   Separate domains for frontend and backend can be set using
```

   :ref:`$TYPO3_CONF_VARS[BE][cookieDomain]<typo3ConfVars_be_cookieDomain>`
   respectively.

.. index::
   TYPO3_CONF_VARS SYS; trustedHostsPattern
.. _typo3ConfVars_sys_trustedHostsPattern:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['trustedHostsPattern']
=========================================================

.. confval:: trustedHostsPattern

   :type: text
   :Default: 'SERVER_NAME'

   Regular expression pattern that matches all allowed hostnames (including
   their ports) of this TYPO3 installation, or the string :php:`SERVER_NAME`
   (default).

   The default value :php:`SERVER_NAME` checks if the HTTP Host header equals
   the SERVER_NAME and SERVER_PORT. This is secure in correctly configured
   hosting environments and does not need further configuration. If you cannot
   change your hosting environment, you can enter a regular expression here.

   Examples:

   :php:`.*\.domain\.com` matches all hosts that end with
   :file:`.domain.com` with all corresponding subdomains.

   :php:`(.*\.domain|.*\.otherdomain)\.com` matches all hostnames with
   subdomains from :file:`.domain.com` and :file:`.otherdomain.com`.

>    Be aware that HTTP Host header may also contain a port. If your installation
> 
>    runs on a specific port, you need to explicitly allow this in your pattern,
> 
>    for example :php:`www\.domain\.com:88` allows only :file:`www.domain.com:88`,
>    **not** :file:`www.domain.com`. To disable this check completely
> 
>    (not recommended because it is **insecure**) you can use :php:`.*` as pattern.

Reviewer: ```suggestion
   Be aware that HTTP Host header may also contain a port. If your installation
   runs on a specific port, you need to explicitly allow this in your pattern,
   for example :php:`www\.domain\.com:88` allows only :file:`www.domain.com:88`,
   **not** :file:`www.domain.com`. To disable this check completely
   (not recommended because it is **insecure**) you can use :php:`.*` as pattern.
```

============================================

.. confval:: ddmmyy

   :type: text
   :Default: 'd-m-y'

   Format of Day-Month-Year - see PHP-function `date() <https//php.net/date>`__

.. index::
   TYPO3_CONF_VARS SYS; hhmm
.. _typo3ConfVars_sys_hhmm:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['hhmm']
==========================================

.. confval:: hhmm

   :type: text
   :Default: 'H:i'

   Format of Hours-Minutes - see PHP-function `date() <https//php.net/date>`__

.. index::
   TYPO3_CONF_VARS SYS; USdateFormat
.. _typo3ConfVars_sys_USdateFormat:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['USdateFormat']
==================================================

.. confval:: USdateFormat

   :type: bool
   :Default: false

   If :php:`TRUE`, dates entered in the TCEforms of the backend will be
   formatted :php:`mm-dd-yyyy`

.. index::
   TYPO3_CONF_VARS SYS; loginCopyrightWarrantyProvider
.. _typo3ConfVars_sys_loginCopyrightWarrantyProvider:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['loginCopyrightWarrantyProvider']
====================================================================

.. confval:: loginCopyrightWarrantyProvider

   :type: text
   :Default: ''

>    If you provide warranty for TYPO3 to your customers insert you (company)

Reviewer: ```suggestion
   If you provide warranty for TYPO3 to your customers, insert your company
```

>    name here. It will appear in the login-dialog as the warranty provider.

Reviewer: ```suggestion
   name here. It will display in the login-dialog as the warranty provider.
```

   (You must also set URL below).

.. index::
   TYPO3_CONF_VARS SYS; loginCopyrightWarrantyURL
.. _typo3ConfVars_sys_loginCopyrightWarrantyURL:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['loginCopyrightWarrantyURL']
===============================================================

.. confval:: loginCopyrightWarrantyURL

   :type: text
   :Default: ''

>    Add the URL where you explain the extend of the warranty you provide.

Reviewer: ```suggestion
   Add the URL where you explain the extent of the warranty you provide.
```

   This URL is displayed in the login dialog as the place where people can
   learn more about the conditions of your warranty. Must be set
   (more than 10 chars) in addition with the
   :ref:`loginCopyrightWarrantyProvider<typo3ConfVars_sys_loginCopyrightWarrantyProvider>`
   message.

.. index::
   TYPO3_CONF_VARS SYS; textfile_ext
.. _typo3ConfVars_sys_textfile_ext:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['textfile_ext']
==================================================

.. confval:: textfile_ext

   :type: text
   :Default: 'txt,ts,typoscript,html,htm,css,tmpl,js,sql,xml,csv,xlf,yaml,yml'

   Text file extensions. Those that can be edited. Executable PHP files may not
   be editable if disallowed!

.. index::
   TYPO3_CONF_VARS SYS; mediafile_ext
.. _typo3ConfVars_sys_mediafile_ext:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['mediafile_ext']
===================================================

.. confval:: mediafile_ext

   :type: text
   :Default: 'gif,jpg,jpeg,bmp,png,pdf,svg,ai,mp3,wav,mp4,ogg,flac,opus,webm,youtube,vimeo'

>    Commalist of file extensions perceived as media files by TYPO3.

Reviewer: ```suggestion
   Comma-list of file extensions perceived as media files by TYPO3.
```

   Must be written in lower case with no spaces between.

.. index::
   TYPO3_CONF_VARS SYS; binPath
.. _typo3ConfVars_sys_binPath:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['binPath']
=============================================

.. confval:: binPath

   :type: text
   :Default: ''

   List of absolute paths where external programs should be searched for.

>    for example :php:`/usr/local/webbin/,/home/xyz/bin/`. (ImageMagick path have to

Reviewer: ```suggestion
   for example :php:`/usr/local/webbin/,/home/xyz/bin/`. (ImageMagick path has to
```

   be configured separately)

.. index::
   TYPO3_CONF_VARS SYS; binSetup
.. _typo3ConfVars_sys_binSetup:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['binSetup']
==============================================

.. confval:: binSetup

   :type: multiline
   :Default: ''

   List of programs (separated by newline or comma). By default programs
   will be searched in default paths and the special paths defined by

>    :ref:`binPath<typo3ConfVars_sys_binPath>`. When PHP has :php:`openbasedir`
>    enabled the programs can not be found and have to be configured here.

Reviewer: ```suggestion
   :ref:`binPath<typo3ConfVars_sys_binPath>`. When PHP has :php:`openbasedir`
   enabled, the programs can not be found and have to be configured here.
```

   Example: :php:`perl=/usr/bin/perl,unzip=/usr/local/bin/unzip`

.. index::
   TYPO3_CONF_VARS SYS; setMemoryLimit
.. _typo3ConfVars_sys_setMemoryLimit:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['setMemoryLimit']
====================================================

.. confval:: setMemoryLimit

   :type: int
   :Default: 0

>    Memory limit in MB: If more than 16, TYPO3 will try to use :php:`ini_set()`

Reviewer: what is MB?

Author: mega byte. But it is usally shortened like that.

         :ref:`[SYS][productionExceptionHandler]<typo3ConfVars_sys_productionExceptionHandler>`
         is used as exception handler.

      1
         Debug: Display error messages with the registered
         :ref:`[SYS][errorHandler]<typo3ConfVars_sys_errorHandler>`.
         Sets :php:`display_errors=1`. The configured
         :ref:`[SYS][debugExceptionHandler]<typo3ConfVars_sys_debugExceptionHandler>`
         is used as exception handler.

   Configures whether PHP errors or Exceptions should be displayed,
   effectively setting the PHP option :php:`display_errors` during runtime.

.. index::
   TYPO3_CONF_VARS SYS; productionExceptionHandler
.. _typo3ConfVars_sys_productionExceptionHandler:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['productionExceptionHandler']
================================================================

.. confval:: productionExceptionHandler

   :type: phpClass
   :Default: :php:`\TYPO3\CMS\Core\Error\ProductionExceptionHandler::class`

   Classname to handle exceptions that might happen in the TYPO3-code. Leave
   this empty to disable exception handling.  The default exception handler displays

>    a nice error message when something went wrong. The error message is

Reviewer: ```suggestion
   a nice error message when something goes wrong. The error message is
```

   anonymization task only, not while creating new log entries.

.. index::
   TYPO3_CONF_VARS SYS; systemMaintainers
.. _typo3ConfVars_sys_systemMaintainers:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['systemMaintainers']
=======================================================

.. confval:: systemMaintainers

   :type: array
   :Default: null

   A list of backend user IDs allowed to access the Install Tool

.. index::
   TYPO3_CONF_VARS SYS; features
.. _typo3ConfVars_sys_features:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['features']
==============================================

New features of TYPO3 that are activated on new installations but upgrading
installations may still use the old behaviour.

.. index::
   TYPO3_CONF_VARS SYS; features form.legacyUploadMimeTypes
.. _typo3ConfVars_sys_features_form.legacyUploadMimeTypes:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['features']['form.legacyUploadMimeTypes']
----------------------------------------------------------------------------

.. confval:: form.legacyUploadMimeTypes:

   :type: bool
   :Default: true

   If on, some mime types are predefined for the "FileUpload" and "ImageUpload"

>    elements of the "form" extension which always allows file uploads of these

Reviewer: ```suggestion
   elements of the "form" extension, which always allows file uploads of these
```

   counted and last hit time is logged to the database.

.. index::
   TYPO3_CONF_VARS SYS; features security.backend.enforceReferrer
.. _typo3ConfVars_sys_features_security.backend.enforceReferrer:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['features']['security.backend.enforceReferrer']
----------------------------------------------------------------------------------

.. confval:: security.backend.enforceReferrer

   :type: bool
   :Default: true

   If on, HTTP referrer headers are enforced for backend and install tool requests to mitigate
   potential same-site request forgery attacks. The behavior can be disabled in case HTTP proxies filter
   required referer header. As this is a potential security risk, it is recommended to enable this option.

.. index::
   TYPO3_CONF_VARS SYS; features yamlImportsFollowDeclarationOrder
.. _typo3ConfVars_sys_features_yamlImportsFollowDeclarationOrder:

$GLOBALS['TYPO3_CONF_VARS']['SYS']['features']['yamlImportsFollowDeclarationOrder']
-----------------------------------------------------------------------------------

.. confval:: yamlImportsFollowDeclarationOrder

   :type: bool
   :Default: false

>    If on, the YAML imports are imported in the order they are defined in the importing YAML configuration.'

Reviewer: ```suggestion
   If on, the YAML imports are imported in the order they are defined in the importing YAML configuration.
```
